### PR TITLE
Real time support for GRPC

### PR DESCRIPTION
**Pull request by KhushalPShah**

### CODE REVIEW

File: utils/client.go [hunk 1]
```diff
@@ -51,16 +93,38 @@ func (c *Client) Close() {
...
+				return
+			}
+			var data map[string]interface{}
+			data["Token"] = in.Token
+			data["DBType"] = in.DbType
+			data["Project"] = in.Project
+			data["Group"] = in.Group
+			data["Type"] = in.Type
+			data["ID"] = in.Id
+			data["Where"] = in.Where
```
Comment (YourTechBud) [inline utils/client.go]: i guess `in.Where` is a []byte. The `data["where"]` must be a map[string]interface{} i guess. You should Unmarshal it first.

File: utils/client.go [hunk 2]
```diff
@@ -70,8 +134,16 @@ func (c *Client) ClientID() string {
 
 // CreateWebsocketClient makes a client object to manage the socket
 func CreateWebsocketClient(socket *websocket.Conn) *Client {
-	channel := make(chan interface{}, 5)
+	channel := make(chan *model.Message, 5)
+	ctx, cancel := context.WithCancel(context.Background())
+	id := uuid.NewV1().String()
+	return &Client{id, channel, ctx, cancel, socket, nil, "WEBSOCKET"}
```
Comment (YourTechBud) [inline utils/client.go]: can we make the `WEBSOCKET` a constant in the utils package?

File: utils/client.go [hunk 3]
```diff
@@ -70,8 +134,16 @@ func (c *Client) ClientID() string {
...
+}
+
+// CreateGRPCClient makes a client object to manage the grpc
+func CreateGRPCClient(stream pb.SpaceCloud_RealTimeServer) *Client {
+	channel := make(chan *model.Message, 5)
 	ctx, cancel := context.WithCancel(context.Background())
 	id := uuid.NewV1().String()
-	return &Client{id, channel, ctx, cancel, socket}
+	return &Client{id, channel, ctx, cancel, nil, stream, "GRPC"}
```
Comment (YourTechBud) [inline utils/client.go]: can we make the `GRPC` a constant in the utils package.

File: utils/client.go [hunk 4]
```diff
@@ -51,16 +93,38 @@ func (c *Client) Close() {
...
 // Read startes a blocking reader routine
 func (c *Client) Read(cb DataCallback) {
-	defer c.Close()
-	for {
-		data := &model.Message{}
-		err := c.socket.ReadJSON(data)
-		if err != nil {
-			return
+	if c.protocol == "WEBSOCKET" {
+		defer c.Close()
```
Comment (YourTechBud) [inline utils/client.go]: can we put differ on the top just once?

File: utils/client.go [hunk 5]
```diff
@@ -2,40 +2,82 @@ package utils
...
+					grpcResponse := pb.RealTimeResponse{Id: res.ID, Group: responseMsg.Group, Ack: responseMsg.Ack, Error: responseMsg.Error, FeedData: feedData}
+					c.stream.Send(&grpcResponse)
+
+				case TypeRealtimeFeed:
+					feed := res.Data.(model.FeedData)
+					var feedData []*pb.FeedData
+					payload, err := json.Marshal(feed.Payload)
+					if err != nil {
+						grpcResponse := pb.RealTimeResponse{Error: err.Error()}
+						c.stream.Send(&grpcResponse)
```
Comment (YourTechBud) [inline utils/client.go]: Just log the error. Dont stream errors to the client. Also you need to return straight after logging the error.

File: utils/client.go [hunk 6]
```diff
@@ -2,40 +2,82 @@ package utils
...
+							c.stream.Send(&grpcResponse)
+						}
+						feedData[i] = &pb.FeedData{QueryId: feed.QueryID, DocId: feed.DocID, Type: feed.Type, Group: feed.Group, DbType: feed.DBType, Payload: payload, TimeStamp: feed.TimeStamp}
+					}
+					grpcResponse := pb.RealTimeResponse{Id: res.ID, Group: responseMsg.Group, Ack: responseMsg.Ack, Error: responseMsg.Error, FeedData: feedData}
+					c.stream.Send(&grpcResponse)
+
+				case TypeRealtimeFeed:
+					feed := res.Data.(model.FeedData)
+					var feedData []*pb.FeedData
```
Comment (YourTechBud) [inline utils/client.go]: feedData should be defined and declared using something like
```go
feedData := make([]*pb.FeedData, 1)
```

Comment (YourTechBud) [reply utils/client.go]:
```suggestion
					feedData := make([]*pb.FeedData, 1)
```

File: utils/client.go [hunk 7]
```diff
@@ -52,15 +94,41 @@ func (c *Client) Close() {
 // Read startes a blocking reader routine
 func (c *Client) Read(cb DataCallback) {
 	defer c.Close()
-	for {
-		data := &model.Message{}
-		err := c.socket.ReadJSON(data)
-		if err != nil {
-			return
+	if c.protocol == Websocket {
```
Comment (YourTechBud) [inline utils/client.go]: Can we use a switch case here?

File: utils/client.go [hunk 8]
```diff
@@ -2,40 +2,82 @@ package utils
...
 type DataCallback func(data *model.Message)
 
 // RoutineWrite starts a json writer routine
 func (c *Client) RoutineWrite() {
-	go func() {
-		for res := range c.channel {
-			err := c.socket.WriteJSON(res)
-			if err != nil {
-				log.Println(err)
+	if c.protocol == Websocket {
```
Comment (YourTechBud) [inline utils/client.go]: Can we use a switch case here

File: utils/client.go [hunk 9]
```diff
@@ -2,40 +2,83 @@ package utils
...
-	}()
+		}()
+	case GRPC:
+		go func() {
+			for res := range c.channel {
+				//Convert the Message into RealTime response.
+				switch res.Type {
+				case TypeRealtimeSubscribe, TypeRealtimeUnsubscribe:
+					//Decode the Message
+					feedData := make([]*pb.FeedData, 1)
```
Comment (YourTechBud) [inline utils/client.go]:
```suggestion
					feedData := make([]*pb.FeedData, len(responseMsg.Docs))
```

File: utils/client.go [hunk 10]
```diff
@@ -52,14 +95,42 @@ func (c *Client) Close() {
...
+			data["Token"] = in.Token
+			data["DBType"] = in.DbType
+			data["Project"] = in.Project
+			data["Group"] = in.Group
+			data["Type"] = in.Type
+			data["ID"] = in.Id
+			var temp interface{}
+			err = json.Unmarshal(in.Where, &temp)
+			if err != nil {
+				return
```
Comment (YourTechBud) [inline utils/client.go]: Let's log the error here as well!

File: utils/client.go (hunk 9 above)
Comment (YourTechBud) [reply utils/client.go]: you'll have to move line number 49 above 48

**Comment by YourTechBud** [comment]
Also, you've got to resolve the merge conflict as well!

File: utils/client.go (hunk 9 above)
Comment (KhushalPShah) [reply utils/client.go]: How can I get the Length of the Docs?
Cos that will correspondingly be the Length of my array of pb.FeedData.

Comment (KhushalPShah) [reply utils/client.go]: Got it..Thanks!

**Pull request merged**